message-broker/src/lib.rs

```diff
@@ -127,53 +126,154 @@ decl_module! {
+			// than the announced, we would waste some of weight. In the case the actual value is
+			// greater than the announced, we will miss opportunity to send a couple of messages.
+			weight += T::DbWeight::get().reads_writes(1, 1);
+			let hrmp_max_message_num_per_candidate =
+				<cumulus_parachain_upgrade::Module<T>>::host_configuration()
+					.map(|cfg| cfg.hrmp_max_message_num_per_candidate)
+					.unwrap_or(0);
+			AnnouncedHrmpMessagesPerCandidate::put(hrmp_max_message_num_per_candidate);
+
+			// NOTE that the actual weight consumed by `on_finalize` may turn out lower.
 			weight += T::DbWeight::get().reads_writes(
-				2 + HRMP_MSG_NUM_PER_CANDIDATE as u64,
-				4 + HRMP_MSG_NUM_PER_CANDIDATE as u64,
+				3 + hrmp_max_message_num_per_candidate as u64,
+				4 + hrmp_max_message_num_per_candidate as u64,
 			);
 
 			weight
 		}
 
 		fn on_finalize() {
+			let host_config = <cumulus_parachain_upgrade::Module<T>>::host_configuration().unwrap();
+			let relevant_messaging_state = <cumulus_parachain_upgrade::Module<T>>::relevant_messaging_state().unwrap();
+
 			<Self as Store>::PendingUpwardMessages::mutate(|up| {
-				let num = cmp::min(UMP_MSG_NUM_PER_CANDIDATE, up.len());
-				storage::unhashed::put(
-					well_known_keys::UPWARD_MESSAGES,
-					&up[0..num],
+				let (count, size) = relevant_messaging_state.relay_dispatch_queue_size;
+
+				let available_capacity = cmp::min(
+					host_config.max_upward_queue_count.saturating_sub(count),
+					host_config.max_upward_message_num_per_candidate,
 				);
+				let available_size = host_config.max_upward_queue_size.saturating_sub(size);
+
+				// Count the number of messages we can possibly fit in the given constraints, i.e.
+				// available_capacity and available_size.
+				let num = up
+					.iter()
+					.scan(
+						(available_capacity as usize, available_size as usize),
+						|(cnt, size), msg| match (cnt.checked_sub(1), size.checked_sub(msg.len())) {
+							(Some(cnt), Some(size)) => Some((cnt, size)),
+							_ => None,
+						},
+					)
+					.count();
+
+				// TODO: #274 Return back messages that do not longer fit into the queue.
+
+				storage::unhashed::put(well_known_keys::UPWARD_MESSAGES, &up[0..num]);
 				*up = up.split_off(num);
 			});
 
-			// Sending HRMP messages is a little bit more involved. On top of the number of messages
-			// per block limit, there is also a constraint that it's possible to send only a single
-			// message to a given recipient per candidate.
+			// Sending HRMP messages is a little bit more involved. There are the following
+			// constraints:
+			//
+			// - a channel should exist (and it can be closed while a message is buffered),
+			// - at most one message can be sent in a channel,
+			// - the sent out messages should be ordered by ascension of recipient para id.
+			// - the capacity and total size of the channel is limited,
+			// - the maximum size of a message is limited (and can potentially be changed),
```

Comment:
Now here is the comment correct, why do we allow to change the max message limit at all? :D

Reply:
Even funnier thing is that it's not even possible to renegotiate the limits for an existing channel.

For now it's hard to see how the cross chain messaging will turn out and how exactly it will be used. For now, it seems to me that allow the renegotiation of the sizes may come in handy for the parachains due to various reasons.

They off course can shut down the channel and open another one, but that would take some latency. Since we already have this possibility that the messages may be returned to the sender for some reason, we can use it as well for this case. So it doesn't really add a lot of extra complexity.